src/trans/gpu/external/setup_trans.F90

```diff
@@ -470,49 +468,56 @@ SUBROUTINE SETUP_TRANS(KSMAX,KDGL,KDLON,KLOEN,LDSPLIT,PSTRET,&
 
 ! Initialize A arrays
 
-ALLOCATE(FG%ZAA(ALIGN(R%NDGNH,8),ALIGN((R%NTMAX+2)/2,8),D%NUMP))
-ALLOCATE(FG%ZAS(ALIGN(R%NDGNH,8),ALIGN((R%NTMAX+3)/2,8),D%NUMP))
+ALLOCATE(FG%ZAA(D%OFFSETS_GEMM_MATRIX(D%NUMP+1)))
+ALLOCATE(FG%ZAS(D%OFFSETS_GEMM_MATRIX(D%NUMP+1)))
 
-FG%ZAA(:,:,:) = 0._JPRBT
-FG%ZAS(:,:,:) = 0._JPRBT
+FG%ZAA(:) = 0._JPRBT
+FG%ZAS(:) = 0._JPRBT
 
+IMLOC0 = 0
 DO JMLOC=1,D%NUMP
   KM = D%MYMS(JMLOC)
   KDGLU = G%NDGLU(KM)
   ILA = (R%NSMAX-KM+2)/2
   ILS = (R%NSMAX-KM+3)/2
 
-  FG%ZAA(1:KDGLU,1:ILA,JMLOC)=S%FA(JMLOC)%RPNMA(1:KDGLU,1:ILA)
-  FG%ZAS(1:KDGLU,1:ILS,JMLOC)=S%FA(JMLOC)%RPNMS(1:KDGLU,1:ILS)
-ENDDO
+  IF (KM /= 0) THEN
+    CALL C_F_POINTER(C_LOC(FG%ZAA(1+D%OFFSETS_GEMM_MATRIX(JMLOC))), LOCAL_ARR, &
+        & (/D%LEGENDRE_MATRIX_STRIDES(JMLOC),ILA/))
+    LOCAL_ARR(1:KDGLU,1:ILA) = S%FA(JMLOC)%RPNMA(1:KDGLU,1:ILA)
 
-! arrays for m=0 in ledir_mod:
-IMLOC0 = FINDLOC(D%MYMS,0)
-IF(IMLOC0(1) > 0) THEN
-  ALLOCATE(FG%ZAA0(SIZE(FG%ZAA,1),SIZE(FG%ZAA,2)))
-  ALLOCATE(FG%ZAS0(SIZE(FG%ZAS,1),SIZE(FG%ZAS,2)))
-  FG%ZAA0 = FG%ZAA(:,:,IMLOC0(1))
-  FG%ZAS0 = FG%ZAS(:,:,IMLOC0(1))
-ENDIF
+    CALL C_F_POINTER(C_LOC(FG%ZAS(1+D%OFFSETS_GEMM_MATRIX(JMLOC))), LOCAL_ARR, &
+        & (/D%LEGENDRE_MATRIX_STRIDES(JMLOC),ILS/))
+    LOCAL_ARR(1:KDGLU,1:ILS) = S%FA(JMLOC)%RPNMS(1:KDGLU,1:ILS)
+  ELSE
+    IMLOC0 = JMLOC
+    ALLOCATE(FG%ZAA0(ALIGN(KDGLU,8),ILA))
+    ALLOCATE(FG%ZAS0(ALIGN(KDGLU,8),ILS))
+
+    FG%ZAA0(:,:) = 0
+    FG%ZAS0(:,:) = 0
+    FG%ZAA0(1:KDGLU,1:ILA)=S%FA(JMLOC)%RPNMA(1:KDGLU,1:ILA)
+    FG%ZAS0(1:KDGLU,1:ILS)=S%FA(JMLOC)%RPNMS(1:KDGLU,1:ILS)
+  ENDIF
+ENDDO
 
 ALLOCATE(FG%ZEPSNM(D%NUMP,0:R%NTMAX+2))
 FG%ZEPSNM = 0._JPRBT
-CALL PREPSNM !Initialize on the host
-
+CALL PREPSNM
 WRITE(NOUT,*)'setup_trans: sizes1 NUMP=',D%NUMP
 #ifdef ACCGPU
   WRITE(NOUT,*) 'Using OpenACC'
 #endif
 #ifdef OMPGPU
   WRITE(NOUT,*) 'Using OpenMP offloading'
 #endif
-WRITE(NOUT,'(A10,":",I9,"B")') 'FG%ZAS', C_SIZEOF(FG%ZAS(1,1,1))*SIZE(FG%ZAS)
-WRITE(NOUT,'(A10,":",I9,"B")') 'FG%ZAA', C_SIZEOF(FG%ZAA(1,1,1))*SIZE(FG%ZAA)
-WRITE(NOUT,'(A10,":",I9,"B")') 'FG%ZAS0', C_SIZEOF(FG%ZAS0(1,1))*SIZE(FG%ZAS0)
-WRITE(NOUT,'(A10,":",I9,"B")') 'FG%ZAA0', C_SIZEOF(FG%ZAA0(1,1))*SIZE(FG%ZAA0)
-WRITE(NOUT,'(A10,":",I9,"B")') 'FG%ZEPSNM', C_SIZEOF(FG%ZEPSNM(1,1))*SIZE(FG%ZEPSNM)
+WRITE(NOUT,'(A10,":",I11,"B")') 'FG%ZAS', C_SIZEOF(FG%ZAS(1))*SIZE(FG%ZAS)
+WRITE(NOUT,'(A10,":",I11,"B")') 'FG%ZAA', C_SIZEOF(FG%ZAA(1))*SIZE(FG%ZAA)
+WRITE(NOUT,'(A10,":",I11,"B")') 'FG%ZAS0', C_SIZEOF(FG%ZAS0(1,1))*SIZE(FG%ZAS0)
+WRITE(NOUT,'(A10,":",I11,"B")') 'FG%ZAA0', C_SIZEOF(FG%ZAA0(1,1))*SIZE(FG%ZAA0)
+WRITE(NOUT,'(A10,":",I11,"B")') 'FG%ZEPSNM', C_SIZEOF(FG%ZEPSNM(1,1))*SIZE(FG%ZEPSNM)
 
-IF (IMLOC0(1) > 0) THEN
+IF (IMLOC0 > 0) THEN
```

Comment:
If I understand correctly, `KMLOC` has the role of "are any of the m's on this rank equal to 0?" Why not make this more transparent, by deleting `IMLOC0`, and replacing this if statement with
```
IF (ANY(D%MYMS == 0)) THEN
```
What do you think?

Reply:
Yes, exactly, it tries to see if there is a m==0 (because those are computed in double precision). In other places I do `IMLOC0 = FINDLOC(D%MYMS,0)` because oftentimes we need the actual position of M==0, but agree in setup_trans we can use the way you propose. I will change this. (IMLOC0 used to be a global variable KMLOC0 until very recently when I removed it to enable various resolutions).